### DISCUSSION TITLE
Memory optimization for single node cluster

### DISCUSSION THREAD

**Discussion by DgS4dd0n** [k3s-io/k3s#10961] [General]
From what i understand, k3s its a single binary that loads in RAM and decompress in RAM all needed components.
I remember something about unload unused components from RAM at start time but i can't find it.
Second option is to build the binary without the components that i don't need (traefik, servicelb, metrics-server, coredns, dashboard, kube-proxy, network-policy, etcd-snapshots, agent, flannel, helm-controller, scheduler). Can anyone please help me with that.

**Comment by dereknola** [comment]
If you don't want to use all the components of K3s, you can always utilize the `--disable` flag. This isn't going to save you that much memory, but you can certainly try. https://docs.k3s.io/installation/packaged-components?_highlight=disab#disabling-manifests

**Reply by DgS4dd0n** [reply]
Thanks, but let's say i want to rebuild it without metrics-server, helm-controller, ... I removed traefik from image-list.txt
What are the components that i can remove with a flag in build command?

**Reply by brandond** [reply]
That's not how it works. Preventing stuff from being built into the binary won't save you any memory at runtime. If you don't want it, just disable it.

**Reply by DgS4dd0n** [reply]
From what i understood, k3s it's a single compressed binary that load itself into RAM, decompress itself into RAM with all it's components (in case you need them). They are not started with --disable, but are loaded and use RAM.

**Reply by brandond** [reply]
The bits that are extracted to disk are not kept in memory after they're extracted out. By the time the system is up and running the Kubernetes components, whatever memory might have temporarily been used for extracting the few KB of manifests and binaries has already been freed and re-used for other things. You're not going to see any change from removing things from the build.
